### ISSUE TITLE
[Emmet] Include should be more like DocumentSelector

### ISSUE THREAD

**Issue by KamasamaK** [Microsoft/vscode#49588]
Currently, the Emmet extension has a setting called `emmet.includeLanguages` which just targets language IDs to map to the Emmet-capable language ID. I think it would be very useful if instead it was more like `DocumentSelector`, which also allows [`DocumentFilter`](https://code.visualstudio.com/docs/extensionAPI/vscode-api#DocumentFilter) instead of just language ID.

**Comment by ramya-rao-a** [comment]
In what scenario would this be useful?

**Comment by KamasamaK** [comment]
The use case I was considering is for the language CFML, which I use. Even though .cfm and .cfc files utilize the same language, they have different purposes. I find it more of an annoyance to have Emmet enabled for .cfc files (basically class files) and would prefer that it only be for .cfm files (view templates). The ability to use the `pattern` in `DocumentFilter` would be able to accomplish this while still allowing language ID to be used.

More generally, there are probably many use cases that I cannot think of and are not relevant or known to me that brought about the decision to use the current form of `DocumentSelector` for registering providers that would be just as applicable.

**Comment by ramya-rao-a** [comment]
Supporting the complete `DocumentFilter` in the `emmet.includeLanguages` setting gets complicated. Instead, supporting just the pattern part of the DocumentFilter should be doable.

So in your case, that would look like
```
"emmet.includeLanguages": {
    "*.cfm": "html"
}
```

PRs are welcome.

All references to the [getMappingForIncludedLanguages](https://github.com/Microsoft/vscode/search?utf8=✓&q=getMappingForIncludedLanguages&type=) method will need to be updated. 

On a side note,

I was looking at https://marketplace.visualstudio.com/items?itemName=KamasamaK.vscode-cfml and found a setting `cfml.emmet.enable`. What does this do and how does it tie to the `emmet.includeLanguages` setting?

**Comment by KamasamaK** [comment]
I realize supporting the complete `DocumentFilter` in the `emmet.includeLanguages` would not work since the key is the issue and keys cannot be complex types, but I was trying to make the request irrespective of implementation details and figured someone would come up with a clever solution. I would be hesitant to overload the key in the way suggested if `*` and `.` are legal characters in a language ID.

The setting `cfml.emmet.enable`, which is disabled by default, adds `"cfml": "html"` to the user's `emmet.includeLanguages` setting. It was just something I wanted to see if I could do as I was implementing the extension to help the user get up-and-running quicker, but I will probably remove that setting eventually and just add instructions instead since it is not difficult to add yourself and I would rather minimize setting things for other extensions.

**Comment by ramya-rao-a** [comment]
> I would be hesitant to overload the key in the way suggested if * and . are legal characters in a language ID

They are indeed legal characters. But that need not stop us from supporting [GlobPattern](https://github.com/Microsoft/vscode/blob/49170e54b123f9daa01afbb6d2e058cd81bcbfa4/src/vs/vscode.d.ts#L1810) as the key in `emmet.includeLanguages`. 

If current file matches both a language id and a glob pattern, the mapping for the glob pattern can win. Thoughts?

**Comment by KamasamaK** [comment]
I've glanced at the extension source you linked and have a decent understanding of these parts of the API, so correct me if I'm mistaken, but I'm not sure how the "current file" would be taken into consideration unless it re-registers the `CompletionProvider` in an event like `onDidChangeActiveTextEditor`. Registration looks to currently only be done on extension activation and `onDidChangeConfiguration`.

I would say checking `vscode.languages.getLanguages()` should be sufficient at the time of registration to determine whether it be interpreted as a language ID or `GlobPattern`. The only issue there being if they install/enable an extension after startup. I don't have a strong opinion on which should be prioritized in the case of a conflict, but I would lean to language ID.

**Comment by ramya-rao-a** [comment]
good point on the fact that current file doesnt help with completion provider registration.

> The only issue there being if they install/enable an extension after that activation that contributes a new language

Installing a new extension requires reload of VS Code window which activates the emmet extension from scratch ending up registering the completion providers again. So we are good there

**Comment by KamasamaK** [comment]
@ramya-rao-a I haven't had time to really look at this, but if I do get around to it would you still be willing to accept this as a PR? It doesn't seem too difficult at first glance.
